packages/flutter/lib/src/services/undo_manager.dart

```diff
@@ -0,0 +1,111 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+
+import '../../services.dart';
+
+/// A low-level interface to the system's undo manager.
+///
+/// To receive events from the system UndoManager, create an
+/// [UndoManagerClient] and set it as the [client] on [UndoManager].
+///
+/// The [setUndoState] method can be used to update the system's UndoManager
+/// using the [canUndo] and [canRedo] parameters.
+///
+/// When the system undo or redo button is tapped, the current
+/// [UndoManagerClient] will receive [UndoManagerClient.handlePlatformUndo]
+/// with a string representing whether the event is "undo" or "redo".
+///
+/// Currently, only iOS has an UndoManagerPlugin implemented on the engine side.
+///
+/// See also:
+///
+/// * [NSUndoManager](https://developer.apple.com/documentation/foundation/nsundomanager)
+class UndoManager {
+  UndoManager._() {
+    _channel = SystemChannels.undoManager;
+    _channel.setMethodCallHandler(_handleUndoManagerInvocation);
+  }
+
+  /// Set the [MethodChannel] used to communicate with the system's text input
+  /// control.
+  ///
+  /// This is only meant for testing within the Flutter SDK. Changing this
+  /// will break the ability to input text. This has no effect if asserts are
+  /// disabled.
```

Comment:
This sounds like TextInputClient, can you tweak them so they're about undo/redo?